### PR TITLE
synthesis: minimal synthesis to drive MAX_UNGROUP_SIZE policy

### PR DESCRIPTION
**Pull request by oharboe**
@gadfort @maliberty Thoughts?

This roughly cuts hierarchical synthesis times in half.

### CODE REVIEW

**Comment by gadfort** [comment]
@oharboe what are you attempting to accomplish with this?

**Comment by oharboe** [comment]
> @oharboe what are you attempting to accomplish with this?

The first stage of synthesis takes seconds instead of hours this way...

**Comment by gadfort** [comment]
Instead of relying on an area estimate (I've been using a instance count instead, which doesn't require a techmap to occur). Running two sets of synthesis is wasteful (I don't think the current setup in ORFS is doing us any favors and this just pushes off fixing it properly)

**Comment by oharboe** [comment]
> Instead of relying on an area estimate (I've been using a instance count instead, which doesn't require a techmap to occur). Running two sets of synthesis is wasteful (I don't think the current setup in ORFS is doing us any favors and this just pushes off fixing it properly)

I'm counting cells instead of area in this PR.

**Comment by gadfort** [comment]
I see, still without running a techmap it's not clear that you are getting a proper count since $lcu and other complex structures would still be left behind. If your method works properly, I don't see the need for the python script since then you can just have a single synthesis run without the need to invoke yosys twice.

**Comment by oharboe** [comment]
> I see, still without running a techmap it's not clear that you are getting a proper count since $lcu and other complex structures would still be left behind. If your method works properly, I don't see the need for the python script since then you can just have a single synthesis run without the need to invoke yosys twice.

Good point, that would be much nicer.

How do you "use instance count"? Is that in the SiliconCompiler?

**Comment by gadfort** [comment]
you can see https://github.com/siliconcompiler/siliconcompiler/blob/1122fadb92b4ca5803b7b66f0b071f3185310ffb/siliconcompiler/tools/yosys/syn_asic.tcl#L185
Like @maliberty said this has not been evaluated to make sure it works in all cases, but I suspect it would work for you, except it will take longer than a few seconds since it's doing some extra work along the way.

**Comment by oharboe** [comment]
> you can see https://github.com/siliconcompiler/siliconcompiler/blob/1122fadb92b4ca5803b7b66f0b071f3185310ffb/siliconcompiler/tools/yosys/syn_asic.tcl#L185 Like @maliberty said this has not been evaluated to make sure it works in all cases, but I suspect it would work for you, except it will take longer than a few seconds since it's doing some extra work along the way.

Interesting. How is this .tcl file run? It looks like you running it from tclsh and are importing yosys as a module and restarting synthesis a number of times, so there's some global state kept in the yosys module across "yosys foo" calls.

**Comment by gadfort** [comment]
it's run via SC, we don't import yosys commands so the tcl script is purely tcl and yosys calls are via the `yosys foo` (which is how yosys set it up). It runs several trial flattenings to see when the design is no longer "benefits" from flattening. Long term yosys needs a better system which is compiled in since that would give better access to the FFs etc which is really what determines the level of flattening that is possible.

**Comment by oharboe** [comment]
> it's run via SC, we don't import yosys commands so the tcl script is purely tcl and yosys calls are via the `yosys foo` (which is how yosys set it up). It runs several trial flattenings to see when the design is no longer "benefits" from flattening. Long term yosys needs a better system which is compiled in since that would give better access to the FFs etc which is really what determines the level of flattening that is possible.

Silly question: how is state persisted between invocations of yosys in your .tcl script? In ram? on disk?

**Comment by gadfort** [comment]
in ram, the call https://github.com/siliconcompiler/siliconcompiler/blob/1122fadb92b4ca5803b7b66f0b071f3185310ffb/siliconcompiler/tools/yosys/syn_asic.tcl#L41
saves the state (that way I can make some speculative changes and see if they are any good before commiting to that)
and https://github.com/siliconcompiler/siliconcompiler/blob/1122fadb92b4ca5803b7b66f0b071f3185310ffb/siliconcompiler/tools/yosys/syn_asic.tcl#L56
loads it back into the working memory.

**Comment by oharboe** [comment]
@gadfort @maliberty Single pass hierarchical synthesis. The first pass is very fast. Use number of cells as driver for the policy on whether to flatten modules or not.

There is a failure, but I wouldn't expect identical results.

I tried it on MegaBoom and it ran through rtlmp, which fails for a flattened design.

This implementation is a mix between tcl & python. The tcl interface in Yosys leaves something to be desired, so there's something to be said for exporting a .json file of everything and using Python or Tcl to examine the model, rather than try to use the yosys interface...

**Comment by oharboe** [comment]
@maliberty Trying to graph the new policy vs. proper cell count after full synthesis when using cells for MAX_UNGROUP_SIZE.

Note! This only shows whether the policy is similar, it doesn't show anything about the difference in quality of results!

```
make DESIGN_CONFIG=designs/nangate45/bp_multi_top/config.mk synth
```

NOTE!!! I fixed the label, the red stars is "flattened in neither case", i.e. policy is the same.

image

Adding to fast build:

```
techmap
opt -fast -full -purge
```

image

**Comment by oharboe** [comment]
@maliberty Cell count and area are not correlated in an obvious way... my thinking is that the hieararchy is not taken into account. A module that has almost no cells could contain two modules that has lots of cells... In that case, which modules do you keep, which do you flatten?

I'm puzzled by the current policy. Is there a reason it was done this way or was this just the way it was implemented initially and as it turned out it gave better numbers than the triton partitioner?

image

**Comment by oharboe** [comment]
Using `stat -tech cmos` we get an estimate of number of transistors which correlate strongly to area after very coarse synthesis.

image

**Comment by oharboe** [comment]
The policy is very nearly identical...

image

File: flow/scripts/plot_cells.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+with open('objects/nangate45/bp_multi/base/synth.json', 'r') as file:
+    data_b = json.load(file)
+
+# Get the cell counts for each module in each file
+# create dictionary of module names and their area, strip the \ in front of module name
+# Area of module \bsg_mem_1r1w_synth_width_p99_els_p2_read_write_same_addr_p0_harden_p0 is 1821.834000
+with open('reports/nangate45/bp_multi/base/synth_cmos.txt', 'r') as file:
+    cmos = json.load(file)
+    transistors = {name.replace('\\', ''):
+                   int(module['estimated_num_transistors'].replace('+', ''))
```
Comment (maliberty) [inline flow/scripts/plot_cells.py:20]: Note that the '+' means there were gates that couldn't be counted and were assigned zero.

**Pull request closed without merging**